Issue#589 Support MERGE INTO ... USING syntax

stumc opened the pull request:
Support standard MERGE statement: http://en.wikipedia.org/wiki/Merge_%28SQL%29
(Duplicated from http://www.h2database.com/html/roadmap.html)

The current MERGE implementation is non-standard as it is missing the USING clause.

Database management systems Oracle Database, DB2, Teradata, EXASOL, Firebird, CUBRID, HSQLDB, MS SQL, Vectorwise and Apache Derby & Postgres support the standard syntax of the SQL 2003/2008 MERGE command:
```
 MERGE INTO targetTable AS T USING sourceTable AS S ON (T.ID = S.ID)
   WHEN MATCHED THEN
       UPDATE SET column1 = value1 [, column2 = value2 ...] WHERE column1=valueUpdate
       DELETE WHERE column1=valueDelete
   WHEN NOT MATCHED THEN
       INSERT (column1 [, column2 ...]) VALUES (value1 [, value2 ...]);
```
Only Oracle support the additional optional DELETE clause. 
The current implementation (for comparison) uses this syntax:
```
 MERGE INTO tablename [(columnName1,...)] 
 [KEY (keyColumnName1,...)]
 [ VALUES (expression1,...) | SELECT ...]
```
This fix will add the standard SQL command syntax.

This change is all my own work - I own the rights - and I offer it unrestricted to h2database project.

stumc committed:
Issue#589 Support standard MERGE USING statement

stumc committed:
Merge branch 'master' of https://github.com/h2database/h2database into Issue#589

stumc committed:
Merge branch 'master' of https://github.com/h2database/h2database into Issue#589

stumc committed:
Merge branch 'Issue#589' of https://github.com/stumc/h2database into Issue#589

stumc committed:
Asking for help at this point - not for merging

stumc committed:
Basic empty table case works with lots of debug.

stumc committed:
Joined onCOndition and fix for bad named select column alias

stumc committed:
Huge headway in testing

stumc committed:
Finished set of test cases

stumc committed:
Added docs

stumc committed:
Added extra test case for multiple updates

stumc committed:
Added optional AS alias to docs

stumc committed:
Refined extra error messages

stumc committed:
Removed debug statements

stumc committed:
Code review checks

stumc committed:
Merge branch 'master' into Issue#589

stumc committed:
Workaround for null schemaName issue

stumc committed:
TestPgServer failing unless currentToken test was done in toUpperCase()

stumc committed:
Added readIfAll() to cope with all-or-nothing parsing for WHEN MATCHED THEN/WHEN NOT MATCHED THEN syntax

stumc committed:
Added comment for new method parseIfAll()

stumc committed:
Added Oracle case dual update and delete allowed on same row

stumc committed:
Bypassed null schemaName bug with special parseSelect() call

grandinj commented:
thanks a lot for this work, just to let you know it will take a while to find a big enough block of time to review this

stumc commented:
@grandinj  Yeah, it's pretty big - and not quite finished yet. I have three cleanup items to do - but this pull request is more for me to objectively view it before getting you and/or others involved. I'll add another comment when this is ready for the big time. No worries about speed on this one.

stumc committed:
Reverted to simple change for schemaName null bug

stumc committed:
Separated Merge and MergeUsing heirarchy

stumc committed:
Missed TestAll.java update

stumc committed:
Finally a schemaName issue solution

stumc committed:
Cleanup of MergeUsing useless stuff copied from Merge

stumc committed:
Removed unused schemaName revert in pasreSelect()

grandinj commented on h2/src/main/org/h2/command/Parser.java:
```diff
@@ -698,6 +700,13 @@ private Schema getSchema(String schemaName) {
+    
+    private Schema getSchemaWithDefault() {
+        if(schemaName==null){
+            schemaName = session.getCurrentSchemaName();
```
is it wise to update this state variable here? If we have multiple commands separated by ";", will this potentially affect following commands

grandinj commented on h2/src/main/org/h2/command/Parser.java:
```diff
@@ -812,6 +825,23 @@ private TableFilter readSimpleTableFilter(int orderInFrom) {
+        } else if (currentTokenType == IDENTIFIER) {
+            String upperCaseCurrentToken = currentToken.toUpperCase();
+            if (!equalsToken("SET", upperCaseCurrentToken) && !excludeTokens.contains(upperCaseCurrentToken)) {
+                System.out.println("currentToken="+currentToken);
```
println

grandinj commented on h2/src/main/org/h2/command/Parser.java:
```diff
@@ -812,6 +825,23 @@ private TableFilter readSimpleTableFilter(int orderInFrom) {
+        if (readIf("AS")) {
+            alias = readAliasIdentifier();
+        } else if (currentTokenType == IDENTIFIER) {
+            String upperCaseCurrentToken = currentToken.toUpperCase();
```
I think you need an enhanced equalsToken() method because of the "only do uppercasing in certain situations" down there

grandinj commented on h2/src/main/org/h2/command/Parser.java:
```diff
@@ -1317,18 +1439,21 @@ private IndexHints parseIndexHints(Table table) {
-            // left and right are not keywords (because they are functions as
-            // well)
-            if (!isToken("LEFT") && !isToken("RIGHT") && !isToken("FULL")) {
+        } else if (currentTokenType == IDENTIFIER && !excludeIdentifiers.contains(currentToken)) {
```
I think you need an enhanced equalsToken() method because of the uppercasing down there

grandinj commented on h2/src/main/org/h2/command/Parser.java:
```diff
@@ -3311,7 +3448,28 @@ private boolean readIf(String token) {
+     * Reads every token in list, in order - returns true if all are found.
+     * If any are not found, returns false - AND resets parsing back to state when called.
+     */
+    private boolean readIfAll(List<String> tokens) {
```
might be nice to use varargs here, ie.
    readIfAll(String... tokens)

grandinj commented on h2/src/main/org/h2/command/Parser.java:
```diff
@@ -3311,7 +3448,28 @@ private boolean readIf(String token) {
+        // save parse location in case we have to fail this test
+        int start = lastParseIndex;
+        for(String token: tokens){
+            String currentTokenUpper = currentToken.toUpperCase();
```
shouldn't need to uppercase here, equalsToken() already does that

grandinj commented on h2/src/main/org/h2/command/Parser.java:
```diff
@@ -5021,50 +5179,66 @@ private TableView parseSingleCommonTableExpression() {
+        return view;
+    }
+
+    private List<Column> createQueryColumnTemplateList(String[] cols, Query withQuery, String[] querySQLOutput) {
```
please add comment that you are using querySQLOutput String[] as trick to get an "out" parameter

(I like to document anything "interesting" to make maintenance easier)

grandinj commented on h2/src/main/org/h2/command/Parser.java:
```diff
@@ -5021,50 +5179,66 @@ private TableView parseSingleCommonTableExpression() {
                 parameters, columnTemplateList.toArray(new Column[0]), session,
-                true/* recursive */, false);
-        if (!view.isRecursiveQueryDetected()) {
+                allowRecursiveQueryDetection/* recursive */, false);
```
can remove the 
   /* recursive */
commnent now, I put that in for boolean params when passing a literal, to make it more obvious what the param is for

grandinj commented on h2/src/main/org/h2/command/dml/Delete.java:
```diff
@@ -30,30 +30,35 @@
 
     private Expression condition;
-    private TableFilter tableFilter;
+    private TableFilter targetTableFilter;
```
please add a comment like // for MERGE..USING

stumc replied:
done

grandinj commented on h2/src/main/org/h2/command/dml/Merge.java:
```diff
@@ -33,10 +33,11 @@
 public class Merge extends Prepared {
 
-    private Table table;
+    private Table targetTable;
```
why does Merge change if you are creating a MergeUsing class?

stumc replied:
I renamed two fields and the getters/setters - to clarify their use - no other modifications made. 
As it happens - I learnt a lot from this class because I based the new merge in it - a great deal.

grandinj commented on h2/src/main/org/h2/command/dml/Insert.java:
```diff
@@ -46,6 +47,7 @@
     private boolean sortedInsertMode;
     private int rowNumber;
     private boolean insertFromSelect;
+    private TableFilter sourceTableFilter;
```
please add a comment like // for MERGE..USING

stumc replied:
done

grandinj reviewed (changes requested):
Looks pretty good, some commetns and some questions

stumc committed:
Code review updates for grandinj

stumc commented on h2/src/main/org/h2/command/Parser.java:
```diff
@@ -698,6 +701,32 @@ private Schema getSchema(String schemaName) {
     private Schema getSchema() {
         return getSchema(schemaName);
     }
+    /*
```
@grandinj - I added an explanation in the code instead of reply to your review comment - please read this. (Sorry I have limited internet right now - so this was the best choice on with the few megabytes of internet I had today...)

If you like - I can remove the side effect of setting the new schemaName, returning it only - but the comments still apply...

stumc commented on h2/src/main/org/h2/command/dml/MergeUsing.java:
```diff
@@ -0,0 +1,496 @@
+import org.h2.util.StatementBuilder;
+import org.h2.value.Value;
+
+/**
```
@grandinj Please read this lengthy module comment - it explains a lot about this class and why it implemented like this... also one potential issue mentioned at the end (aka a future bug report)

stumc replied:
Added fix for potential bug report

stumc commented:
@grandinj I think I addressed all of your comments... (even if I didn't reply to each one in the review individually - check the code for increased documentation.

stumc committed:
Added match-select query for ROWID checking

stumc committed:
Added trigger testing and removed source key uniqueness test

PascalSchumacher commented:
It's great to have this feature added to h2! Thank you very much!
